kafka-connect-s3/src/test/java/io/confluent/connect/s3/DataWriterParquetTest.java

```diff
@@ -506,6 +532,205 @@ public void testProjectNoVersion() throws Exception {
     }
   }
 
+  class SchemaConfig {
+    public String name;
+    public boolean optionalItems = false;
+    public boolean regularItems = false;
+    public boolean mapRegular = false;
+    public boolean mapOptional = false;
+    public SchemaConfig nested = null;
+    public SchemaConfig nestedArray = null;
+
+    public SchemaConfig() {}
+
+    public SchemaConfig(
+        String name,
+        boolean regularItems,
+        boolean optionalItems,
+        boolean mapRegular,
+        boolean mapOptional,
+        SchemaConfig nested,
+        SchemaConfig nestedArray
+    ) {
+      this.name = name;
+      this.optionalItems = optionalItems;
+      this.regularItems = regularItems;
+      this.mapRegular = mapRegular;
+      this.mapOptional = mapOptional;
+      this.nested = nested;
+      this.nestedArray = nestedArray;
+    }
+
+    public boolean hasOptionalItems() {
+      if (optionalItems || mapOptional) {
+        return true;
+      }
+      if (nested != null && nested.hasOptionalItems()) {
+        return true;
+      }
+      return nestedArray != null && nestedArray.hasOptionalItems();
+    }
+
+    private Schema create() {
+      SchemaBuilder builder = SchemaBuilder.struct();
+      if (StringUtils.isNotBlank(name)) {
+        builder.name(name);
+      }
+      if (regularItems) {
+        builder.field("regular_items", SchemaBuilder.array(Schema.STRING_SCHEMA).build());
+      }
+      if (optionalItems) {
+        builder.field("optional_items", SchemaBuilder.array(Schema.OPTIONAL_STRING_SCHEMA).build());
+      }
+      if (mapRegular) {
+        builder.field("regular_map", SchemaBuilder.map(
+            Schema.STRING_SCHEMA,
+            SchemaBuilder.array(Schema.STRING_SCHEMA).build()
+        ).build());
+      }
+      if (mapOptional) {
+        builder.field("optional_map", SchemaBuilder.map(
+            Schema.STRING_SCHEMA,
+            SchemaBuilder.array(Schema.OPTIONAL_STRING_SCHEMA).build()
+        ).build());
+      }
+      if (this.nested != null) {
+        builder.field("nested", nested.create());
+      }
+      if (this.nestedArray != null) {
+        builder.field("nested_array", SchemaBuilder.array(nestedArray.create()));
+      }
+      return builder.build();
+    }
+
+  }
+
+  /**
+   * Tests ParquetRecordWriterProvider::schemaHasArrayOfOptionalItems()
+   *
+   * Test permutations of schemas and schema nesting with and without optional array items
+   * somewhere within the schema.
+   */
+  @Test
+  public void testSchemaHasArrayOfOptionalItems() {
+    for (int regularItems = 0; regularItems < 2; ++regularItems) {
+      for (int optionalItems = 0; optionalItems < 2; ++optionalItems) {
+        for (int mapRegular = 0; mapRegular < 2; ++mapRegular) {
+          for (int mapOptional = 0; mapOptional < 2; ++mapOptional) {
+            for (int has_nested = 0; has_nested < 2; ++has_nested) {
```

Comment:
Could we please fix these 2 variables also `has_nested` and `has_nested_array` ?

Reply:
weird, done.